### DISCUSSION TITLE
Annotate the Resource class

### DISCUSSION THREAD

**Discussion by fowork1** [thecodingmachine/graphqlite#480] [Q&A]
How i should annotate resource class extends standard JsonResource?

in part of code ProductController it look like:
```php
/**
* @param int $id
* @return ProductResource
* @query()
*/
public function testProduct(int $id): ProductResource
{
    $product = Product::findOrFail($id);
    return new ProductResource($product);
}
```

and ProductResource:
```php
/**
 * Class ProductResource
 * @property Product $product
 * @package App\Http\Resources
 *
 * @Type
 * @MagicField(name="array", outputType="array")
 */
final class ProductResource extends JsonResource
{
    use ProductResourceTrait;

    /**
     * @Field
     * @Autowire(for="$request")
     * @param Request $request
     * @return array
     */
    public function toArray($request): array
    {
        $data = [];
        $outfit = $this->getOutfit();

        if (!empty($outfit)) {
            $data['outfit'] = $outfit;
        } else {
            $data['outfit'] = null;
        }

        return array_merge(
            $this->getBaseInfo(),
            [
                'variations' => $this->getVariations(),
                'info' => $this->getExtraInfo(),
            ],
            $data
        );
    }
```

on trying i got error
```
For parameter $request, in App\\Http\\Resources\\ProductResource::toArray, 
cannot map class \"Illuminate\\Http\\Request\" to a known GraphQL input type. 
Check your TypeMapper configuration.
```

I read about @Factory but i supposing that's not my case.
Also i can't use @input (@UseInputType) annotation for Request class...
Is exist solution for this case?

**Comment by oojacoboo** [comment]
@fowork1 I'm guessing you're going to need to write a custom root type mapper to handle the Request object.  GraphQLite is simply saying that you must have the `Illuminate\Http\Request` class defined as a type.  It doesn't know how to handle it.

I'm really not sure what you're trying to do here though.  It looks wrong on many levels.  I don't know why you need this Laravel mess or why you're trying to output an array like this.  This looks like some REST design.

At any rate, take a look at the follow.  Maybe it's what you want.
https://graphqlite.thecodingmachine.io/docs/4.1/custom-types#registering-a-custom-scalar-type-advanced

IMO though, I think you're doing it all wrong and making a mistake with your approach.

**Reply by fowork1** [reply]
Thanks for your answer!
That is the REST design )
It's because i trying implement GraphQL on existing project to make a proof of concept. I want to make an AB test for REST vs GraphQL without large effort. That's why GraphQLite.
If result will be positive then i'll migrate whole project to GraphQL.

> IMO though, I think you're doing it all wrong and making a mistake with your approach.

I would be extremely grateful if you could advise me on a more effective approach. (I hope it won't require rewriting a tangible part of the project or duplicating business logic)

**Reply by oojacoboo** [reply]
Just implement it according to the docs and don't try to do all this array casting and array output stuff.  Add the proper `Type` and `Field` attributes/annotations.  You'll annotate your getters as fields on the class and you'll remove all this array magic.
